app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js

```diff
@@ -0,0 +1,90 @@
+function insertAfter(newNode, existingNode) {
+    existingNode.parentNode.insertBefore(newNode, existingNode.nextSibling);
+}
+
+function removeAllChildNodes(parent) {
+    while (parent.firstChild) {
+        parent.removeChild(parent.firstChild);
+    }
+}
+
+function showOrHideVisualizationDiv(divId, jsonString) {
+    if (jsonString != '' && jsonString != 'null') {
+            document.getElementById(divId).style.display = 'block';
+        } else {
+            document.getElementById(divId).style.display = 'none';
+    }
+}
+
+function getGridDimensions(json) {
+    let allX = [];
+    let allY = [];
+    let allW = [];
+    let allH = [];
+
+    for (let i = 0; i < json.length; i++) {
+        json[i].x == undefined ? undefined : allX.push(json[i].x);
+        json[i].y == undefined ? undefined : allY.push(json[i].y);
+        json[i].w == undefined ? undefined : allW.push(json[i].w);
+        json[i].h == undefined ? undefined : allH.push(json[i].h);
+    }
+    let totalWidth = (allX.length != json.length | allW.length != json.length) ? json.length : Math.max(...allX.map(function(num, idx) {return num+allW[idx]}));
+    let totalHeight = (allY.length != json.length | allH.length != json.length) ? 1 : Math.max(...allY.map(function(num, idx) {return num+allH[idx]}));
+
+    return [totalHeight, totalWidth]
+}
+
+function createViewportDiv(divId, viewportNum, viewportSpec, totalHeight, totalWidth) {
+    let viewportDiv = document.createElement("div");
+    viewportDiv.setAttribute('id', viewportSpec.viewport_name);
+    viewportDiv.style.opacity = '0.5';
+    viewportDiv.style.position = 'absolute';
+    viewportDiv.style.zIndex = '-' + viewportSpec.order;
+    viewportDiv.style.width = isNaN(viewportSpec.w) ? (1 / parseFloat(totalWidth).toFixed(2))*100 + '%' : (viewportSpec.w / parseFloat(totalWidth).toFixed(2))*100 + '%';

```

Comment:
The `toFixed()` turns a Number into a string with a certain number of decimals. And javascript tries to change the type of a certain variable so that the operation applied to it makes sense (type coercion). So what will happen when the interpreter comes across this statement (notice the quotes to indicate that it is a string):
```
> const totalWidth = "123.456"
undefined
> parseFloat(totalWidth)
123.456
> parseFloat(totalWidth).toFixed(2)
'123.46'
> 1 / parseFloat(totalWidth).toFixed(2)
0.008099789405475458
> (1 / parseFloat(totalWidth).toFixed(2)) * 100
0.8099789405475458
> (1 / parseFloat(totalWidth).toFixed(2)) * 100 + '%'
'0.8099789405475458%'
```

So that is probably not quite what you wanted to do. I think you wanted to do something like this:
```
(1 / parseFloat(totalWidth) * 100).toFixed(2) + '%'
```
And even here the `parseFloat` is not strictly necessary because the interpreter will already try to convert the string `totalWidth` into a Number when you try to apply the divide operation on it, but it is clearer to add it.